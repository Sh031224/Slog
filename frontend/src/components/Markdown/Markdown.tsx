import React, { useEffect } from "react";
import MarkDown from "markdown-to-jsx";
import "./Markdown.scss";
import Prism from "lib/prism/prism";

interface MarkdownProps {
  children: string;
}

const markedH1 = ({ children }: { children: string }) => {
  return <h1 id={children.toString().replace(/\ /g, "-")}>{children}</h1>;
};

const markedH2 = ({ children }: { children: string }) => {
  return <h2 id={children.toString().replace(/\ /g, "-")}>{children}</h2>;
};

const markedH3 = ({ children }: { children: string }) => {
  return <h3 id={children.toString().replace(/\ /g, "-")}>{children}</h3>;
};

const markedH4 = ({ children }: { children: string }) => {
  return <h4 id={children.toString().replace(/\ /g, "-")}>{children}</h4>;
};

const markedH5 = ({ children }: { children: string }) => {
  return <h5 id={children.toString().replace(/\ /g, "-")}>{children}</h5>;
};

const markedH6 = ({ children }: { children: string }) => {
  return <h6 id={children.toString().replace(/\ /g, "-")}>{children}</h6>;
};

const markedA = ({ children, ...props }: { children: string }) => {
  return (
    <a {...props} target="_blank">
      {children}
    </a>
  );
};

const Markdown = ({ children }: MarkdownProps) => {
  useEffect(() => {
    Prism.highlightAll();
  }, [children]);

  return (
    <div className="markdown-code">
      <MarkDown
        options={{
          overrides: {
            h1: markedH1,
            h2: markedH2,
            h3: markedH3,
            h4: markedH4,
            h5: markedH5,
            h6: markedH6,
            a: markedA
          },
          forceBlock: true
        }}
      >
        {children}
      </MarkDown>
    </div>
  );
};

export default React.memo(Markdown);
