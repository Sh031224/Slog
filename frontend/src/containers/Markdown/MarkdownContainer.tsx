import React from "react";
import dynamic from "next/dynamic";

const Markdown = dynamic(() => import("components/Markdown"));

interface MarkdownContainerProps {
  className?: string;
  children: string;
}

const MarkdownContainer = ({ children, className }: MarkdownContainerProps) => {
  return (
    <div className={className}>
      <Markdown>{typeof children === "string" ? children : ""}</Markdown>
    </div>
  );
};

export default React.memo(MarkdownContainer);
