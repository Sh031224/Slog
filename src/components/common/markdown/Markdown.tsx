import MarkdownJSX from "markdown-to-jsx";
import { Prism as SyntaxHighlighter } from "react-syntax-highlighter";
import styled from "styled-components";
import { dracula } from "react-syntax-highlighter/dist/cjs/styles/prism";

type Props = {
  content: string;
};

const Markdown: React.FC<Props> = ({ content }) => {
  return (
    <MarkdownContainer>
      <MarkdownJSX
        options={{
          overrides: {
            h1: CustomMarkdownH1,
            h2: CustomMarkdownH2,
            h3: CustomMarkdownH3,
            h4: CustomMarkdownH4,
            h5: CustomMarkdownH5,
            h6: CustomMarkdownH6,
            a: CustomMarkdownA,
            code: CustomMarkdownCode
          },
          forceBlock: true
        }}
      >
        {content}
      </MarkdownJSX>
    </MarkdownContainer>
  );
};

const MarkdownContainer = styled.div`
  font-size: 1.125rem;
  color: ${({ theme }) => theme.colors.ftMarkdownBlack};
  line-height: 1.7;
  letter-spacing: -0.004em;
  word-break: keep-all;
  overflow-wrap: break-word;
  ${({ theme }) => theme.device.tablet} {
    font-size: 1rem;
  }
  p + a,
  a {
    color: ${({ theme }) => theme.colors.ftBlue};
    &:hover {
      text-decoration: underline;
    }
  }
  p + h1,
  p + h2,
  p + h3,
  p + h4 {
    margin-top: 2.5rem;
    ${({ theme }) => theme.device.tablet} {
      margin-top: 2rem;
    }
  }
  h1,
  h2,
  h3,
  h4 {
    line-height: 1.5;
    margin-bottom: 1rem;
    ${({ theme }) => theme.device.tablet} {
      margin-bottom: 0.75rem;
    }
  }
  h1 {
    font-size: 2.5rem;
    ${({ theme }) => theme.device.tablet} {
      font-size: 2.25rem;
    }
  }
  h2 {
    font-size: 2rem;
    ${({ theme }) => theme.device.tablet} {
      font-size: 1.75rem;
    }
  }
  h3 {
    font-size: 1.5rem;
    ${({ theme }) => theme.device.tablet} {
      font-size: 1.25rem;
    }
  }
  h4 {
    font-size: 1.125rem;
    ${({ theme }) => theme.device.tablet} {
      font-size: 1rem;
    }
  }
  hr {
    height: 1px;
    width: 100%;
    margin-top: 2rem;
    margin-bottom: 2rem;
    border-width: initial;
    border-style: none;
    border-color: initial;
    border-image: initial;
    background: ${({ theme }) => theme.colors.bgUnderLineWhite};
  }
  blockquote {
    margin-top: 2rem;
    margin-bottom: 2rem;
    border-top-right-radius: 4px;
    border-bottom-right-radius: 4px;
    margin-left: 0px;
    margin-right: 0px;
    color: ${({ theme }) => theme.colors.ftLightBlack};
    border-left: ${({ theme }) => theme.colors.bdBlue} 4px solid;
    background-color: ${({ theme }) => theme.colors.bgLightGray};
    padding: 1rem 1rem 1rem 2rem;
    & :first-child {
      margin-top: 0px;
    }
    & :last-child {
      margin-bottom: 0px;
    }
  }
  pre {
    & > div {
      font-family: "Fira Mono", source-code-pro, Menlo, Monaco, Consolas, "Courier New", monospace;
      overflow-x: auto;
      letter-spacing: 0px;
      border-radius: 5px;
      line-height: 1.5 !important;
      overflow-x: auto !important;
      letter-spacing: 0px !important;
      padding: 1rem !important;
      border-radius: 5px !important;
      font-size: 0.875rem;
      ${({ theme }) => theme.device.tablet} {
        font-size: 0.75rem;
      }
      & > code {
        background: none !important;
      }
    }
  }
  :not(pre) > code[class*="language-"],
  pre[class*="language-"] {
    background: none !important;
  }
  code {
    font-family: "Fira Mono", source-code-pro, Menlo, Monaco, Consolas, "Courier New", monospace;
  }
  p > code {
    background: ${({ theme }) => theme.colors.bgCode};
    padding: 0.2em 0.4em;
    border-radius: 5px;
    font-size: 85%;
  }
  img {
    max-width: 100%;
    height: auto;
    display: block;
    margin-top: 1.5rem;
    margin-bottom: 1.5rem;
  }
  p > img {
    display: block;
    max-width: 100%;
    margin: 3rem auto;
  }
  table {
    border-collapse: collapse;
    display: block;
    width: 100%;
    width: -webkit-max-content;
    width: -moz-max-content;
    width: max-content;
    max-width: 100%;
    overflow: auto;
    tr:nth-child(2n) {
      background-color: ${({ theme }) => theme.colors.bgTableWhite};
    }
    tr {
      background-color: ${({ theme }) => theme.colors.bgWhite};
      border-top: 1px solid ${({ theme }) => theme.colors.bdTrWhite};
    }
    td,
    th {
      padding: 6px 13px;
      border: 1px solid ${({ theme }) => theme.colors.bdTdWhite};
    }
  }
`;

const CustomMarkdownH1 = ({ children }: { children: string }) => {
  return <h1 id={children.toString().replace(/ /g, "-")}>{children}</h1>;
};

const CustomMarkdownH2 = ({ children }: { children: string }) => {
  return <h2 id={children.toString().replace(/ /g, "-")}>{children}</h2>;
};

const CustomMarkdownH3 = ({ children }: { children: string }) => {
  return <h3 id={children.toString().replace(/ /g, "-")}>{children}</h3>;
};

const CustomMarkdownH4 = ({ children }: { children: string }) => {
  return <h4 id={children.toString().replace(/ /g, "-")}>{children}</h4>;
};

const CustomMarkdownH5 = ({ children }: { children: string }) => {
  return <h5 id={children.toString().replace(/ /g, "-")}>{children}</h5>;
};

const CustomMarkdownH6 = ({ children }: { children: string }) => {
  return <h6 id={children.toString().replace(/ /g, "-")}>{children}</h6>;
};

const CustomMarkdownA = ({ children, ...props }: { children: string }) => {
  return (
    <a {...props} target="_blank">
      {children}
    </a>
  );
};

const CustomMarkdownCode = ({ children, ...props }) => {
  const match = /lang-(\w+)/.exec(props.className || "");
  return match ? (
    <SyntaxHighlighter language={match[1]} PreTag="div" style={dracula} {...props}>
      {children}
    </SyntaxHighlighter>
  ) : (
    <code {...props}>{children}</code>
  );
};

export default Markdown;
