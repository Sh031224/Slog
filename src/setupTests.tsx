import { ThemeProvider } from "styled-components";
import { render } from "@testing-library/react";
import { lightTheme } from "style/theme";
import { store } from "stores";
import { Provider } from "react-redux";
import { NotificationContainer } from "react-notifications";

const Providers = ({ children }) => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={lightTheme}>
        <NotificationContainer />
        {children}
      </ThemeProvider>
    </Provider>
  );
};

const customRender = (ui, options = {}) => render(ui, { wrapper: Providers, ...options });

// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom/extend-expect";
import "jest-styled-components";

(window as any).scrollTo = () => {};

export { customRender as render };
