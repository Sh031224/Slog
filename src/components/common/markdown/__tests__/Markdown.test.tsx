import { render } from "setupTests";
import Markdown, { IMarkdownProps } from "../Markdown";

describe("Markdown", () => {
  const setup = (props = {}) => {
    const initialProps: IMarkdownProps = {
      content: "## Testing"
    };

    return render(<Markdown {...initialProps} {...props} />);
  };

  it("matches snapshot", () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it("renders correctly", () => {
    const { getByText } = setup();

    expect(
      getByText(
        (content, element) =>
          element.tagName.toLowerCase() === "h2" && content.startsWith("Testing")
      )
    );
  });
});
