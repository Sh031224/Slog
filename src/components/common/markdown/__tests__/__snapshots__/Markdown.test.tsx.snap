// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown matches snapshot 1`] = `
.c0 {
  font-size: 1.125rem;
  color: #222426;
  line-height: 1.7;
  -webkit-letter-spacing: -0.004em;
  -moz-letter-spacing: -0.004em;
  -ms-letter-spacing: -0.004em;
  letter-spacing: -0.004em;
  word-break: keep-all;
  overflow-wrap: break-word;
}

.c0 p + a,
.c0 a {
  color: #5373f8;
}

.c0 p + a:hover,
.c0 a:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0 p + h1,
.c0 p + h2,
.c0 p + h3,
.c0 p + h4 {
  margin-top: 2.5rem;
}

.c0 h1,
.c0 h2,
.c0 h3,
.c0 h4 {
  line-height: 1.5;
  margin-bottom: 1rem;
}

.c0 h1 {
  font-size: 2.5rem;
}

.c0 h2 {
  font-size: 2rem;
}

.c0 h3 {
  font-size: 1.5rem;
}

.c0 h4 {
  font-size: 1.125rem;
}

.c0 hr {
  height: 1px;
  width: 100%;
  margin-top: 2rem;
  margin-bottom: 2rem;
  border-width: initial;
  border-style: none;
  border-color: initial;
  border-image: initial;
  background: #dedede;
}

.c0 blockquote {
  margin-top: 2rem;
  margin-bottom: 2rem;
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  margin-left: 0px;
  margin-right: 0px;
  color: rgb(33,37,41);
  border-left: #5373f8 4px solid;
  background-color: rgb(241,243,245);
  padding: 1rem 1rem 1rem 2rem;
}

.c0 blockquote:first-child {
  margin-top: 0px;
}

.c0 blockquote:last-child {
  margin-bottom: 0px;
}

.c0 pre > div {
  font-family: "Fira Mono",source-code-pro,Menlo,Monaco,Consolas,"Courier New",monospace;
  overflow-x: auto;
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  border-radius: 5px;
  line-height: 1.5 !important;
  overflow-x: auto !important;
  -webkit-letter-spacing: 0px !important;
  -moz-letter-spacing: 0px !important;
  -ms-letter-spacing: 0px !important;
  letter-spacing: 0px !important;
  padding: 1rem !important;
  border-radius: 5px !important;
  font-size: 0.875rem;
}

.c0 pre > div > code {
  background: none !important;
}

.c0:not(pre) > code[class*="language-"],
.c0 pre[class*="language-"] {
  background: none !important;
}

.c0 code {
  font-family: "Fira Mono",source-code-pro,Menlo,Monaco,Consolas,"Courier New",monospace;
}

.c0 p > code {
  background: #1b1f231c;
  padding: 0.2em 0.4em;
  border-radius: 5px;
  font-size: 85%;
}

.c0 img {
  max-width: 100%;
  height: auto;
  display: block;
  margin-top: 1.5rem;
  margin-bottom: 1.5rem;
}

.c0 p > img {
  display: block;
  max-width: 100%;
  margin: 3rem auto;
}

.c0 table {
  border-collapse: collapse;
  display: block;
  width: 100%;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  max-width: 100%;
  overflow: auto;
}

.c0 table tr:nth-child(2n) {
  background-color: #f6f8fa;
}

.c0 table tr {
  background-color: #ffff;
  border-top: 1px solid #c6cbd1;
}

.c0 table td,
.c0 table th {
  padding: 6px 13px;
  border: 1px solid #dfe2e5;
}

@media only screen and (max-width:767px) {
  .c0 {
    font-size: 1rem;
  }
}

@media only screen and (max-width:767px) {
  .c0 p + h1,
  .c0 p + h2,
  .c0 p + h3,
  .c0 p + h4 {
    margin-top: 2rem;
  }
}

@media only screen and (max-width:767px) {
  .c0 h1,
  .c0 h2,
  .c0 h3,
  .c0 h4 {
    margin-bottom: 0.75rem;
  }
}

@media only screen and (max-width:767px) {
  .c0 h1 {
    font-size: 2.25rem;
  }
}

@media only screen and (max-width:767px) {
  .c0 h2 {
    font-size: 1.75rem;
  }
}

@media only screen and (max-width:767px) {
  .c0 h3 {
    font-size: 1.25rem;
  }
}

@media only screen and (max-width:767px) {
  .c0 h4 {
    font-size: 1rem;
  }
}

@media only screen and (max-width:767px) {
  .c0 pre > div {
    font-size: 0.75rem;
  }
}

<div>
  <div
    class="notification-container notification-container-empty"
  >
    <div />
  </div>
  <div
    class="c0"
  >
    <h2
      id="Testing"
    >
      Testing
    </h2>
  </div>
</div>
`;
